ssl/ssl_cert_comp.c

```diff
@@ -0,0 +1,484 @@
+    OPENSSL_free(data);
+    OPENSSL_free(ret);
+    return NULL;
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_compressed_data(unsigned char *data, size_t len, size_t orig_len, int alg)
+{
+    return OSSL_COMP_CERT_helper(OPENSSL_memdup(data, len), len, orig_len, alg);
+}
+
+__owur static OSSL_COMP_CERT *OSSL_COMP_CERT_uncompressed_data(unsigned char *data, size_t len, int alg)
+{
+    OSSL_COMP_CERT *ret = NULL;
+    size_t max_length;
+    int comp_length;
+    COMP_METHOD *method;
+    unsigned char *comp_data = NULL;
+    COMP_CTX *comp_ctx = NULL;
+
+    switch (alg) {
+    case TLSEXT_comp_cert_brotli:
+        method = COMP_brotli_oneshot();
+        break;
+    case TLSEXT_comp_cert_zlib:
+        method = COMP_zlib();
+        break;
+    case TLSEXT_comp_cert_zstd:
+        method = COMP_zstd_oneshot();
+        break;
+    default:
+        goto err;
+    }
+
+    if ((max_length = ossl_calculate_comp_expansion(alg, len)) == 0
+          || method == NULL
+          || (comp_ctx = COMP_CTX_new(method)) == NULL
+          || (comp_data = OPENSSL_zalloc(max_length)) == NULL)
+        goto err;
+
+    comp_length = COMP_compress_block(comp_ctx, comp_data, max_length, data, len);
+    if (comp_length <= 0)
+        goto err;
+
+    ret = OSSL_COMP_CERT_helper(comp_data, comp_length, len, alg);
+    comp_data = NULL;
+
+ err:
+    OPENSSL_free(comp_data);
+    COMP_CTX_free(comp_ctx);
+    return ret;
+}
+
+void OSSL_COMP_CERT_free(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (cc == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&cc->references, &i, cc->lock);
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_free(cc->data);
+    CRYPTO_THREAD_lock_free(cc->lock);
+    OPENSSL_free(cc);
+}
+int OSSL_COMP_CERT_up_ref(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&cc->references, &i, cc->lock) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+static int ssl_set_cert_comp_pref(int *prefs, int *algs, size_t len)
+{
+    int j = 0;
+    int i;
+    int found = 0;
+    int already_set[TLSEXT_comp_cert_limit];
+    int tmp_prefs[TLSEXT_comp_cert_limit];
+    int num_algs;
+
+    /* clear all algorithms */
+    if (len == 0 || algs == NULL) {
+        memset(prefs, 0, sizeof(tmp_prefs));
+        return 1;
+    }
+
+    num_algs = len / sizeof(*algs);
+
+    /* This will 0-terminate the array */
+    memset(tmp_prefs, 0, sizeof(tmp_prefs));
+    memset(already_set, 0, sizeof(already_set));
+    /* Include only those algorithms we support, ignoring duplicates and unknowns */
+    for (i = 0; i < num_algs; i++) {
+        switch (algs[i]) {
+        case TLSEXT_comp_cert_brotli:
+        case TLSEXT_comp_cert_zlib:
+        case TLSEXT_comp_cert_zstd:
+            /* Check for enablement */
+            if (!ossl_comp_has_alg(algs[i]))
```

Comment:
Possibly we don't need the switch here. Just confirm `algs[i] != 0` and call `ossl_comp_has_alg()`. If its unrecognised it will fail and we ignore it.